SHA256 implementations based on Intel SHA Extensions

sipa opened the pull request:
Based on #13191.

This adds SHA256 implementations that use Intel's SHA Extension instructions (using intrinsics). This needs GCC 4.9 or Clang 3.4.

In addition to #13191, two extra implementations are provided:
* (a) A variable-length SHA256 implementation using SHA extensions.
* (b) A 2-way 64-byte input double-SHA256 implementation using SHA extensions.

Benchmarks for 9001-element Merkle tree root computation on an AMD Ryzen 1800X system:
* Using generic C++ code (pre-#10821): 6.1ms
* Using SSE4 (master, #10821): 4.6ms
* Using 4-way SSE4 specialized for 64-byte inputs (#13191): 2.8ms
* Using 8-way AVX2 specialized for 64-byte inputs (#13191): 2.1ms
* Using 2-way SHA-NI specialized for 64-byte inputs (this PR): 0.56ms

Benchmarks for 32-byte SHA256 on the same system:
* Using SSE4 (master, #10821): 190ns
* Using SHA-NI (this PR): 53ns

Benchmarks for 1000000-byte SHA256 on the same system:
* Using SSE4 (master, #10821): 2.5ms
* Using SHA-NI (this PR): 0.51ms

kallewoof commented on src/crypto/sha256.cpp:
```diff
@@ -173,22 +484,76 @@ bool SelfTest(TransformType tr) {
+    }
+
+#if defined(ENABLE_SSE41) && !defined(BUILD_BITCOIN_INTERNAL)
+    if (have_sse4) {
```
What about
```C++
        ret = "sse4";
#if defined(ENABLE_SSE41) && !defined(BUILD_BITCOIN_INTERNAL)
        TransformD64_4way = sha256d64_sse41::Transform_4way;
        ret += ",sse41";
#endif
    }
```
?

sipa replied:
Done.

Kick1986 commented:
@sipa

Great works as usual!

I just came cross this thread 

https://github.com/armfazh/flo-shani-aesni/blob/master/README.md

I hope you will have time to have a look about what they did!

promag commented:
Concept ACK, nice numbers.

sipa commented:
Rebased.

@Kick1986 Nice, I'll have a look.

Empact commented:
For clang version, looks like they were added in 3.4, but never noted in the release notes.
Source: went from commit date[1] to release date[2] to file in release version[3].
[1] https://github.com/llvm-mirror/clang/commit/b83f5a77335ca8a5b41ba4e17aa8d4bb6248c1e4#diff-c4f203a0f202bf56c364a657b0aecae9
[2] http://releases.llvm.org/
[3] https://github.com/llvm-mirror/clang/blob/release_34/lib/Headers/shaintrin.h

theuni commented on src/crypto/sha256_shani.cpp:
```diff
@@ -0,0 +1,363 @@
+
+    /* Load state */
+    s0 = _mm_loadu_si128((const __m128i*)s);
+    s1 = _mm_loadu_si128((const __m128i*)(s + 4));
```
I think these could be ```_mm_load_si128``` if s[] was alignas(16).

theuni commented on src/Makefile.am:
```diff
@@ -32,6 +32,7 @@ LIBBITCOIN_UTIL=libbitcoin_util.a
 LIBBITCOIN_CRYPTO=crypto/libbitcoin_crypto.a
 LIBBITCOIN_CRYPTO_SSE41=crypto/libbitcoin_crypto_sse41.a
 LIBBITCOIN_CRYPTO_AVX2=crypto/libbitcoin_crypto_avx2.a
+LIBBITCOIN_CRYPTO_SHANI=crypto/libbitcoin_crypto_shani.a
```
These are starting to get out of hand. I think we should take @TheBlueMatt's suggestion and treat LIBBITCOIN_CRYPTO as a collection of these helpers. That way we can just add ```$(LIBBITCOIN_CRYPTO)``` everywhere, and that will pull in the cpu-specific libs as well. Something like:
```
...
LIBBITCOIN_CRYPTO=crypto/libbitcoin_crypto.a
LIBBITCOIN_CRYPTO_AVX2=crypto/libbitcoin_crypto_avx2.a
LIBBITCOIN_CRYPTO_SHANI=crypto/libbitcoin_crypto_shani.a
LIBBITCOIN_CRYPTO+=$(LIBBITCOIN_CRYPTO_AVX2)
LIBBITCOIN_CRYPTO+=$(LIBBITCOIN_CRYPTO_SHANI)
...
```
Then the cpu-specific ones can be dropped from the LDADDs all over the place. I'm happy to do up a patch on top of this if you'd like.

sipa replied:
@theuni Actually, feel like PRing that as a separate PR, before this one goes in? Then @TheBlueMatt and I can both rebase ours on top of yours.

theuni commented:
concept ACK.

I noticed while testing #13400 that I added a bug, but bitcoind started up fine anyway, due to missing sanity checks for the double/4way/8way hashes. Mind adding those?

DesWurstes commented on src/crypto/sha256_shani.cpp:
```diff
@@ -0,0 +1,363 @@
+
+#include <stdint.h>
+#if defined(_MSC_VER)
+#include <immintrin.h>
```
I believe including immintrin.h is enough for both platforms. x86intrin includes immintrin.h, and immintrin.h includes everything that is needed: https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/immintrin.h

sipa replied:
Thanks, fixed.

sipa commented:
Rebased.

sipa committed:
Use immintrin.h everywhere for intrinsics

sipa commented:
Rebased.

theuni commented on src/crypto/sha256_shani.cpp:
```diff
@@ -0,0 +1,359 @@
+#include <stdint.h>
+#include <immintrin.h>
+
+#include "crypto/common.h"
```
Linter is yelling about include style here.

sipa committed:
[Refactor] CPU feature detection logic for SHA256

sipa committed:
Add SHA256 implementation using using Intel SHA intrinsics

sipa commented:
Rebased after #13471 merge.

Also split up the CPU feature detection logic change into its own commit.

> I noticed while testing #13400 that I added a bug, but bitcoind started up fine anyway, due to missing sanity checks for the double/4way/8way hashes. Mind adding those?

That was addressed in #13438.

theuni commented:
SelfTest() is now missing TransformD64_2way :(

utACK otherwise.

jb55 commented:
Not sure what compelled me to do this, and it's probably overkill... but...

Tested ACK 66b2cf1ccfad545a8ec3f2a854e23f647322bf30 with [100k rounds of quickcheck](https://github.com/jb55/sha256-quickcheck) at various optimization levels, but only with the non-two way transform for now.

DesWurstes commented:
Just a nit from older pull requests: Now that it has a custom CPUID function
https://github.com/bitcoin/bitcoin/blob/d96bdd78307bc5469cb8a4d5ca0e6cbc21fe4073/src/crypto/sha256.cpp#L534-L538
including `<cpuid.h>` is not necessary now:

https://github.com/bitcoin/bitcoin/blob/d96bdd78307bc5469cb8a4d5ca0e6cbc21fe4073/src/crypto/sha256.cpp#L12-L16

Thank you for your awesome contributions!

gmaxwell commented:
ACK

laanwj commented:
utACK 66b2cf1ccfad545a8ec3f2a854e23f647322bf30
tested that build passes on FreeBSD+OpenBSD
